core/src/main/scala/org/apache/spark/sql/catalyst/catalog/TiCompositeSessionCatalog.scala

```diff
@@ -0,0 +1,344 @@
+}
+
+/**
+ * Identical Ti catalog policy.
+ * @param tiContext
+ */
+case class IdentityPolicy(tiContext: TiContext) extends CompositeCatalogPolicy {
+  override val primaryCatalog: SessionCatalog = tiContext.tiConcreteCatalog
+  override val secondaryCatalog: SessionCatalog = tiContext.tiConcreteCatalog
+  override val tiConcreteCatalog: TiSessionCatalog = tiContext.tiConcreteCatalog
+  override val legacyCatalog: SessionCatalog = tiContext.legacyCatalog
+}
+
+/**
+ * Legacy catalog first policy.
+ * @param tiContext
+ */
+case class LegacyFirstPolicy(tiContext: TiContext) extends CompositeCatalogPolicy {
+  override val primaryCatalog: SessionCatalog = tiContext.legacyCatalog
+  override val secondaryCatalog: SessionCatalog = tiContext.tiConcreteCatalog
+  override val tiConcreteCatalog: TiSessionCatalog = tiContext.tiConcreteCatalog
+  override val legacyCatalog: SessionCatalog = tiContext.legacyCatalog
+}
+
+/**
+ * A composition of two catalogs that behaves as a concrete catalog.
+ * @param tiContext
+ */
+class TiCompositeSessionCatalog(tiContext: TiContext)
+    extends SessionCatalog(
+      tiContext.tiConcreteCatalog.externalCatalog,
+      EmptyFunctionRegistry,
+      tiContext.sqlContext.conf
+    )
+    with CompositeCatalogPolicy
+    with TiSessionCatalog {
+  // TODO: configuration for policy choosing.
+  val policy: CompositeCatalogPolicy = LegacyFirstPolicy(tiContext)
+  override val primaryCatalog: SessionCatalog = policy.primaryCatalog
+  override val secondaryCatalog: SessionCatalog = policy.secondaryCatalog
+  override val tiConcreteCatalog: TiSessionCatalog = policy.tiConcreteCatalog
+  override val legacyCatalog: SessionCatalog = policy.legacyCatalog
+
+  // Used to manage catalog change by setting current database.
+  private var currentCatalog: SessionCatalog = primaryCatalog
+
+  // Following are routed to Ti catalog.
+  override def catalogOf(database: Option[String]): Option[SessionCatalog] = synchronized {
+    database
+      .map(db => Seq(primaryCatalog, secondaryCatalog).find(_.databaseExists(db)))
+      .getOrElse(Some(currentCatalog))
+  }
+
+  // Following are handled by composite catalog.
+  override def dropDatabase(db: String, ignoreIfNotExists: Boolean, cascade: Boolean): Unit =
+    catalogOf(Some(db))
+      .getOrElse(if (!ignoreIfNotExists) throw new NoSuchDatabaseException(db) else return )
+      .dropDatabase(db, ignoreIfNotExists, cascade)
+
+  override def alterDatabase(dbDefinition: CatalogDatabase): Unit =
+    catalogOf(Some(dbDefinition.name))
+      .getOrElse(throw new NoSuchDatabaseException(dbDefinition.name))
+      .alterDatabase(dbDefinition)
+
+  override def getDatabaseMetadata(db: String): CatalogDatabase =
+    catalogOf(Some(db))
+      .getOrElse(throw new NoSuchDatabaseException(db))
+      .getDatabaseMetadata(db)
+
+  override def databaseExists(db: String): Boolean =
+    primaryCatalog.databaseExists(db) || secondaryCatalog.databaseExists(db)
```

Comment:
Why here use directly primary and secondary instead of catalogOf?

Reply:
catalogOf determines which catalog to use, it depends on databaseExists to gather the correct result.